### PR TITLE
SF 311 case functionality

### PR DESCRIPTION
**Pull request by alex-wood**
# Summary of Changes
- Added the missing [311 case fields](https://dev.socrata.com/foundry/data.sfgov.org/ktji-gk7t) to the `Sf311Case` model
- Added `create`, `index` and `blocked_bike_lane_cases` actions to the `Sf311CasesController`
- Added the ability to retrieve 311 cases by date range
- Added [the `annotate` gem](https://github.com/ctran/annotate_models) to the Gemfile to make it easier to see which table and columns each model corresponds to
- Added [the `api-pagination` gem](https://github.com/davidcelis/api-pagination) to the Gemfile to make the process of paging through large datasets (i.e., `sf311_cases`) simpler
- Renamed the `Sf311Case` table, model, migration, controller and service files from `sf_311_case` to `sf311_case` so Rails would associate the `Sf311Case` model with the right table (i.e., `ActiveRecord` expects a model named `Sf_311_Case` when the table is named `sf_311_cases`).
- Made lots of minor code restructuring and convention changes that I'm happy to revert if you disagree with them

# Notes and Remaining Work
A database seeding step needs to be added to `/db/Dockerfile` so developers are equipped with case data to work with. I've attempted to export data via the `Shapefile` export format utilized for bikeway networks (https://data.sfgov.org/api/geospatial/ktji-gk7t?method=export&format=Shapefile), but the connection seems to timeout; the cases table contains around 3.2 million rows, so we might need to take a different approach. Another idea is to export the data from an API call like https://data.sfgov.org/resource/ktji-gk7t.csv?service_subtype=Blocking_Bicycle_Lane and import it into the `sf311_cases` table, but I haven't figured out how to get that to work in a script just yet.

Once we have seed 311 case data, we can replace the `index` action with the body of `Sf311CasesController#blocked_bike_lane_cases`. Currently, `Sf311CasesController#index` fetches data directly from 311 instead of the database.

Finally, I haven't tested the newly added `Sf311CaseService.create_case` functionality yet. It's disabled in non-production environments at the moment so it won't cause any issues.

### CODE REVIEW

**Comment by omegabytes** [comment]
Thank you!

File: lanebreach-api/app/services/sf311_case_service.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+require 'soda/client'
+
+module Sf311CaseService
+  extend self
```
Comment (sahilda) [inline lanebreach-api/app/services/sf311_case_service.rb]: This is neat, I didn't know about a module extending self acting as a singleton (which is the route I was going to take).

File: lanebreach-api/app/services/sf311_case_service.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  def get_blocked_bike_lane_cases
+    get_cases(Sf311Case::SERVICE_SUBTYPES[:blocked_bike_lane])
+  end
+
+  def create_case(case_attributes)
+    if Rails.env.production?
+      # TODO: Ensure this call behaves as expected
+      client.post(CASE_DATASET_ID, [case_attributes])
+    else
+      Rails.logger.info("#{self.name}##{__method__} called successfully, actual case creation disabled in this environment (#{Rails.env})")
```
Comment (sahilda) [inline lanebreach-api/app/services/sf311_case_service.rb]: I'm not aware of how to access it but Dale was mentioning that sf actually has a test / dev environment for their 311 case data to which you can make POSTs too. Might be worth looking into.

File: lanebreach-api/lib/tasks/auto_annotate_models.rake [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# NOTE: only doing this in development as some production environments (Heroku)
+# NOTE: are sensitive to local FS writes, and besides -- it's just not proper
+# NOTE: to have a dev-mode tool do its thing in production.
+if Rails.env.development?
+  require 'annotate'
```
Comment (sahilda) [inline lanebreach-api/lib/tasks/auto_annotate_models.rake]: This is a cool gem. I've always resorted to checking out `schema.rb` whenever I needed to check what columns exist.

File: lanebreach-api/app/services/sf311_case_service.rb (hunk 2 above)
Comment (alex-wood) [reply lanebreach-api/app/services/sf311_case_service.rb]: Awesome, I'll chat with Dale about that.

**Pull request merged**